File: sdk/dart/lib/src/verify.dart
import 'dart:convert';
import 'dart:typed_data';

import 'types.dart';

typedef SignatureVerifier = Future<bool> Function(
    String payload, String signatureB64, String pubkeyB64);

String registrationPayload(Registration reg) {
  switch (reg.lastAction) {
    case 'CLAIM':
      return 'CLAIM:${reg.name}:${reg.address}';
    case 'BUY':
      return 'BUY:${reg.name}:${reg.address}';
    case 'UPDATE':
      return 'UPDATE:${reg.name}:${reg.address}:${reg.nonce}';
    case 'DELIST':
      return 'DELIST:${reg.name}:${reg.nonce}';
    case 'RELEASE':
      return 'RELEASE:${reg.name}:${reg.nonce}';
    default:
      return '';
  }
}

String listingPayload(Listing listing) =>
    'LIST:${listing.name}:${listing.price}:${listing.nonce}';

String eventActionPayload(ZnsEvent event) {
  switch (event.action) {
    case 'CLAIM':
      return event.ua != null ? 'CLAIM:${event.name}:${event.ua}' : '';
    case 'BUY':
      return event.ua != null ? 'BUY:${event.name}:${event.ua}' : '';
    case 'LIST':
      return event.price != null && event.nonce != null
          ? 'LIST:${event.name}:${event.price}:${event.nonce}'
          : '';
    case 'DELIST':
      return event.nonce != null ? 'DELIST:${event.name}:${event.nonce}' : '';
    case 'RELEASE':
      return event.nonce != null ? 'RELEASE:${event.name}:${event.nonce}' : '';
    case 'UPDATE':
      return event.ua != null && event.nonce != null
          ? 'UPDATE:${event.name}:${event.ua}:${event.nonce}'
          : '';
    default:
      return '';
  }
}

Future<bool> verifyRegistration(
    Registration reg, String adminPubkey, SignatureVerifier verifier) async {
  final sig = reg.signature;
  if (sig == null) return false;
  final payload = registrationPayload(reg);
  if (payload.isEmpty) return false;
  final pubkey = reg.pubkey ?? adminPubkey;
  return verifier(payload, sig, pubkey);
}

Future<bool> verifyListing(
    Listing listing, String adminPubkey, SignatureVerifier verifier) async {
  final payload = listingPayload(listing);
  return verifier(payload, listing.signature, adminPubkey);
}

Future<bool> verifyEvent(
    ZnsEvent event, String adminPubkey, SignatureVerifier verifier) async {
  final sig = event.signature;
  if (sig == null) return false;
  final payload = eventActionPayload(event);
  if (payload.isEmpty) return false;
  final pubkey = event.pubkey ?? adminPubkey;
  return verifier(payload, sig, pubkey);
}

File: sdk/go/verify.go
package zns

import (
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
)

// VerifySignature checks an Ed25519 signature against a base64-encoded public key.
func VerifySignature(payload, signatureB64, pubkeyB64 string) bool {
	sig, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil || len(sig) != 64 {
		return false
	}
	pk, err := base64.StdEncoding.DecodeString(pubkeyB64)
	if err != nil || len(pk) != 32 {
		return false
	}
	return ed25519.Verify(pk, []byte(payload), sig)
}

// RegistrationPayload reconstructs the signing pre-image from a Registration.
func RegistrationPayload(reg Registration) string {
	switch reg.LastAction {
	case "CLAIM":
		return fmt.Sprintf("CLAIM:%s:%s", reg.Name, reg.Address)
	case "BUY":
		return fmt.Sprintf("BUY:%s:%s", reg.Name, reg.Address)
	case "UPDATE":
		return fmt.Sprintf("UPDATE:%s:%s:%d", reg.Name, reg.Address, reg.Nonce)
	case "DELIST":
		return fmt.Sprintf("DELIST:%s:%d", reg.Name, reg.Nonce)
	case "RELEASE":
		return fmt.Sprintf("RELEASE:%s:%d", reg.Name, reg.Nonce)
	default:
		return ""
	}
}

// ListingPayload reconstructs the signing pre-image from a Listing.
func ListingPayload(listing Listing) string {
	return fmt.Sprintf("LIST:%s:%d:%d", listing.Name, listing.Price, listing.Nonce)
}

// VerifyRegistration verifies a Registration's signature.
// If reg.Pubkey is set, verifies against the sovereign user key;
// otherwise verifies against the admin pubkey.
func VerifyRegistration(reg Registration, adminPubkey string) bool {
	sig := reg.Signature
	if sig == nil || *sig == "" {
		return false
	}
	payload := RegistrationPayload(reg)
	if payload == "" {
		return false
	}
	pubkey := adminPubkey
	if reg.Pubkey != nil && *reg.Pubkey != "" {
		pubkey = *reg.Pubkey
	}
	return VerifySignature(payload, *sig, pubkey)
}

// VerifyListing verifies a Listing's signature against the admin pubkey.
func VerifyListing(listing Listing, adminPubkey string) bool {
	payload := ListingPayload(listing)
	return VerifySignature(payload, listing.Signature, adminPubkey)
}
